modules/stochastic_tools/include/state_sim/model/StateSimObjList.h

```diff
@@ -0,0 +1,277 @@
+ */
+template <class T>
+class StateSimObjList
+{
+  //T _value;
+
+protected:
+  std::unordered_map<int, std::unique_ptr<T>> _item_list;
+  std::unordered_map<std::string, int> _name_to_id;
+  std::unordered_map<int, std::unique_ptr<T>> _deleted;
+
+private:
+  //
+
+public:
+  StateSimObjList();
+  T & add(T * to_add);
+  //T& add<T>(T & to_add);
+  void clear();
+  void deleteAll();
+  void remove(const int item_id);
+  bool contains(const T & to_find);
+  // T * findByName(const std::string name, const bool exception = true);
+  // T * findByName(const std::string name, const bool exception = true) const;
+  int size()
+  {
+    return _item_list.size();
+  };
+  T * operator[](int id);
+  T * operator[](int id) const;
+  T * operator[](std::string id);
+  T * operator[](std::string id) const;
+  //T & iteratorItem(typename std::unordered_map<int, std::unique_ptr<T>>::iterator i) {return *(i->second);};
+
+  void setProcessed(bool value);
+  virtual std::string allProcessed();
+
+  bool loaded;
+
+  typename std::unordered_map<int, std::unique_ptr<T>>::iterator begin() { return _item_list.begin(); };
+  typename std::unordered_map<int, std::unique_ptr<T>>::iterator end()  { return _item_list.end(); };
+};
+
+template <class T>
+StateSimObjList<T>::StateSimObjList()
+  : _item_list(),
+    _name_to_id(),
+    _deleted(),
+    loaded(false)
+{
+}
+
+template <class T>
+T &
+StateSimObjList<T>::add(T * to_add)
+{
+  int id = to_add->id();
+  std::string name = to_add->name();
+
+  typename std::unordered_map<int, std::unique_ptr<T>>::const_iterator got = _item_list.find(to_add->id());
+  if (got == _item_list.end())
+  {
+    //put it in the id and object map
+    _item_list.insert(std::make_pair<int, std::unique_ptr<T>>(to_add->id(), nullptr));
+    _item_list[id].reset(to_add);
+
+    //put the name and id in the name lookup
+    _name_to_id.insert(std::make_pair(name, id));
+    mooseAssert(_name_to_id.size() == _item_list.size(), "Item with the same name but different ID already exists '" + name + "'.");
+  }
+  else
+  {
+    mooseAssert(false, "Item " + name + " can't be added, a second item named '" + got->second->name() + "' has the same id - " + std::to_string(id));
+  }
+
+  return *to_add;
+}
+
+
+template <class T>
+void
+StateSimObjList<T>::clear()
+{
+  _item_list.clear();
+  _name_to_id.clear();
+  _deleted.clear();
+}
+
+template <class T>
+bool
+StateSimObjList<T>::contains(const T & to_find)
+{
+  typename std::unordered_map<int, std::unique_ptr<T>>::const_iterator got = _item_list.find(to_find.id());
+  return (got != _item_list.end());
+}
+
+template <class T>
+void
+StateSimObjList<T>::deleteAll()
+{
+  typedef typename std::unordered_map<int, std::unique_ptr<T>>::const_iterator it_type;
+  for (it_type iterator = _item_list.begin(); iterator != _item_list.end(); iterator++)
+  {
+    T * item = iterator->second;
+    _deleted.insert(std::make_pair(item->id(), item));
```

Comment:
That technique does work for all container types, however on closer inspection this probably isn't one of those places. You have to use a lamba or `std::transform` to do it which is arguably more difficult for people to understand. In general though, look for these kinds of constructs.